plugins/bcm57xx/fu-bcm57xx-recovery-device.c

```diff
@@ -0,0 +1,753 @@
+{
+	BcmRegNVMCommand tmp = { 0 };
+	tmp.bits.Done = 1;
+	return fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+						 REG_NVM_COMMAND, tmp.r32, error);
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_nvram_read (FuBcm57xxRecoveryDevice *self,
+				       guint32 address, guint32 *buf, gsize bufsz,
+				       GError **error)
+{
+	for (guint i = 0; i < bufsz; i++) {
+		BcmRegNVMCommand tmp = { 0 };
+		guint32 val32 = 0;
+		if (!fu_bcm57xx_recovery_device_nvram_clear_done (self, error))
+			return FALSE;
+		if (!fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+						       REG_NVM_ADDR, address, error))
+			return FALSE;
+		tmp.bits.Doit = 1;
+		/* write one word at a time to avoid data corruption issue */
+		tmp.bits.First = TRUE;
+		tmp.bits.Last = TRUE;
+		if (!fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+							   REG_NVM_COMMAND, tmp.r32, error))
+			return FALSE;
+		if (!fu_bcm57xx_recovery_device_nvram_wait_done (self, error)) {
+			g_prefix_error (error, "failed to read @0x%x: ", address);
+			return FALSE;
+		}
+		if (!fu_bcm57xx_recovery_device_bar_read (self, FU_BCM57XX_BAR_DEVICE,
+							  REG_NVM_READ, &val32, error))
+			return FALSE;
+		buf[i] = GUINT32_FROM_BE(val32);
+		address += sizeof(guint32);
+		fu_device_set_progress_full (FU_DEVICE (self), i, bufsz);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_nvram_write (FuBcm57xxRecoveryDevice *self,
+					guint32 address, const guint32 *buf, gsize bufsz,
+					GError **error)
+{
+	for (guint i = 0; i < bufsz; i++) {
+		BcmRegNVMCommand tmp = { 0 };
+		if (!fu_bcm57xx_recovery_device_nvram_clear_done (self, error))
+			return FALSE;
+		if (!fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+							   REG_NVM_WRITE, GUINT32_TO_BE(buf[i]), error))
+			return FALSE;
+		if (!fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+							   REG_NVM_ADDR, address, error))
+			return FALSE;
+		tmp.bits.Wr = 1;
+		tmp.bits.Doit = 1;
+		tmp.bits.First = 1; /* this is really slow but prevents corruption */
+		tmp.bits.Last = 1;
+		if (!fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+							   REG_NVM_COMMAND, tmp.r32, error))
+			return FALSE;
+		if (!fu_bcm57xx_recovery_device_nvram_wait_done (self, error)) {
+			g_prefix_error (error, "failed to read @0x%x: ", address);
+			return FALSE;
+		}
+		address += sizeof(guint32);
+		fu_device_set_progress_full (FU_DEVICE (self), i, bufsz);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_detach (FuDevice *device, GError **error)
+{
+	/* unbind tg3 */
+	return fu_device_unbind_driver (device, error);
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_attach (FuDevice *device, GError **error)
+{
+	/* bind tg3 */
+	return fu_device_bind_driver (device, "pci", "tg3", error);
```

Comment:
Thinking through the scenarios here - what if someone doesn't have tg3 compiled as a module or into their kernel?  This presumably will fail, but I don't think that should actually fail the end of the upgrade / write process.  It probably should be a warning.

Reply:
> but I don't think that should actually fail the end of the upgrade

Agreed, fixed.